### PR TITLE
Project Base Creation

### PR DESCRIPTION
**Pull request by raevangregson**
### Read Me Changes
**What I Hope to Accomplish**
This app will be a copy cat of the [Spotify desktop app](https://open.spotify.com/collection/playlists). At this point I have decided to use the [styled-components library](https://www.styled-components.com/) for styling, and [Redux](https://redux.js.org/) for state management. I am most familiar with creating and using class component based React projects, so for this project I hope to make it purely functional component based as well as using the new [React Hook](https://reactjs.org/docs/hooks-intro.html)'s feature.

I hope to focus on organization - writing thorough and thoughtful PR's where writing out the process, will help me continue to learn even when I think I may already know something. Hopefully, I will be able to learn something new with every PR I put up. I would like to hold myself to high standards for organization of thought, and code - writing neat, readable code with well articulated comments when needed. Code that does well balancing effectiveness, and maintainability.
### File Structure
Screen Shot 2019-10-21 at 2 02 33 PM

Above is the basic file structure I will be using for this project. Starting with the Screen Shot 2019-10-21 at 2 17 25 PM, this folder acts as the **Redux middleware**. The action creators go hand in hand with our Screen Shot 2019-10-21 at 2 05 28 PM folder.

Screen Shot 2019-10-21 at 2 21 12 PM

For every action file there is a coordinating action creator file - this is because they work together. Basic Redux action creators are not required. In the Redux documentation they are referenced as [Async Actions](https://redux.js.org/advanced/async-actions). Basically this is where we can perform any actions, or changes on either the data after we receive it or in some cases before receiving the data. This helps keep separate API call related object separate from our components and away from our reducer files as well.  For now I have just set up very basic `user` files that contain an `updateUser` call - this is an easy action to start with and I can continue adding others as the app grows, rather than trying to assume at this point every action I may need.

### Using Actions with Action Creators
We import the specific action creator function we need into the component that requires it, then we dispatch to whatever API or database we may be using, at this point we can either manipulate the data we receive, or keep it as is. Inside of the action creators we import our actions, and after our data is ready to go we call the particular action to update our state as normal.

**Data Flow:**
**Action Creator** ➛ **Action** ➛ **Reducer**

The last file we have related to `actions` and `actionCreators` is the Screen Shot 2019-10-23 at 9 19 43 AM file. This file is strictly to make sure we are staying consistent. Making sure we are using the correct actions, and action creators that coordinate with one another, as well as making sure we are updating the correct state. This is done by importing the same constant in each coordinating file. 

### Components
The Screen Shot 2019-10-23 at 8 48 24 AM folder is where I will store all of my reusable components. I also have the folder Screen Shot 2019-10-23 at 8 56 20 AM,  where individual views that are unique from one another are stored. These files are the files that will typically be responsible for using the reusable components that are created.

### Configs
Screen Shot 2019-10-23 at 8 58 47 AM

The configs file holds the three different configuration files I am currently using for my redux setup. The first file, Screen Shot 2019-10-23 at 9 02 08 AM is where the redux store is created. This is responsible for creating the object that holds our state and will be used by the redux provider to give our application access to it. It's also used for middleware. Inside of my store configuration, for my middleware I am using a check that will instantiate logging if we are in a development environment. This file is what is taking advantage of my Screen Shot 2019-10-23 at 9 06 22 AM file. Inside of the `logger.js` file we are simply printing the state as it changes to our console, and our `configureStore.js` is using it. The Screen Shot 2019-10-23 at 9 08 20 AM file is another logger, this one is specific to the time it takes for our reducer to update its state. It gives us the total time it takes to run through the life cycle of making any kind of calls to our state. This can help us keep an eye on the performance of our application. 

We are also using `thunk` with Redux, this allows us to return a function instead of an object. It basically gives us the ability to take advantage of the ` dispatch()` function and is what is responsible for turning our application from synchronous to asynchronous. 

### Reducer
Screen Shot 2019-10-23 at 9 13 43 AM

The same way there are coordinating `user.js` files in our `action` and `actionCreator` files, there is typically also a coordinating `user.js` reducer file. This file is responsible for holding the state in this case, particularly for the user. This is what the action and action creator ultimately end up updating and/or retrieving information from. The `rootReducer.js` then takes each of the individual state files and combines them into one large state and in turn that is what is passed to our application with the `store` object using the `provider` wrapper. Keeping the state in separate files and combining them is an easy way to stay organized.

### The Rest
Everything else is pretty straight forward. The Screen Shot 2019-10-23 at 9 30 28 AM is the main page of the application and our Screen Shot 2019-10-23 at 9 31 59 AM file is responsible for instantiating our application. The wrapper around our `App` component is the Redux provider which is giving our application access to the `state`.

In this application I will most likely not be using the `serviceWorker.js`. This file helps with performance when applications may be data heavy. It will make your app load faster because it will cache data instead of loading it each and every time. This is also what is used to give web app's offline capabilities.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -5,7 +5,11 @@
   "dependencies": {
     "react": "^16.10.2",
     "react-dom": "^16.10.2",
-    "react-scripts": "3.2.0"
+    "react-redux": "^7.1.1",
+    "react-scripts": "3.2.0",
+    "redux": "^4.0.4",
+    "redux-thunk": "^2.3.0",
+    "styled-components": "^4.4.0"
```
Comment (mleah) [inline package.json:12]: I'm excited to see how you work with `styled-components`.

File: package-lock 2.json [hunk 2]
```diff
@@ -0,0 +1,14571 @@
+{
```
Comment (mleah) [inline package-lock 2.json]: I haven't seen a `package-lock2.json` file before - can you help me understand why there are two of these files in the project?

**Review by mleah — COMMENTED** [review]
This is a really thorough and well-thought-out PR!  I appreciate the background info that gives insight into your thought process.  The comments throughout are good reminders and explanations of how the app is going to be built out.  I'm excited to see how you grow this project as it becomes more complex. 😄

File: package-lock 2.json (hunk 2 above)
Comment (raevangregson) [reply package-lock 2.json]: Hey Mariah - great catch! This was purely just and accident 😅whoops. I removed and pushed up the changes.

**Pull request merged**